src/contrib/mlir/core/pass/ng_dialect_builder.cpp

```diff
@@ -0,0 +1,541 @@
+{
+    /// NgDialectConversionPass is an MLIR ModulePass Given an nGraph sub-graph, represented as
+    /// CompiledKernel node, it
+    /// translates the graph down to nGraph dialect
+
+    class NgDialectConversionPass : public mlir::ModulePass<NgDialectConversionPass>
+    {
+    public:
+        using TensorList = std::vector<descriptor::Tensor*>;
+        using TypeList = llvm::SmallVector<mlir::Type, 4>;
+
+        NgDialectConversionPass(const ngraph::op::CompiledKernel* compiled_kernel,
+                                mlir::MLIRContext* context,
+                                mlir::OpBuilder* builder,
+                                MLIRCompiler* compiler)
+            : m_compiledKernel(compiled_kernel)
+            , m_context(context)
+            , m_builder(builder)
+            , m_compiler(compiler)
+        {
+        }
+
+        NgDialectConversionPass(const NgDialectConversionPass& obj);
+
+    private:
+        struct TensorInfo
+        {
+            // MLIR values this tensor maps to.
+            mlir::Value* m_value;
+        };
+
+    private:
+        // Converts an nGraph sub-graph to MLIR nGraph dialect.
+        void buildNgDialectModule();
+        void buildNgDialect();
+        void runOnModule() override;
+        // Applies any nGraph dialect optimizations
+        void optimizeNgDialect() { /*TODO: Add Core NG dialect optimizations */}
+
+        mlir::Type getMlirType(const descriptor::Tensor* tensor);
+        mlir::Type getMlirType(const element::Type& type);
+        mlir::Type getMlirType(const ngraph::Node* node);
+
+        TensorInfo getTensorValue(descriptor::Tensor* tensor);
+        void updateTensorValue(descriptor::Tensor* tensor, mlir::Value* value);
+
+        template <typename Op>
+        static mlir::Operation* createOp(NgDialectConversionPass& NgDialectObj,
+                                         const ngraph::Node* ngNode)
+        {
+            throw std::runtime_error("Unimplemented op '" + ngNode->description() +
+                                     "' in MLIR Compiler");
+        }
+
+        // Generic op lowerer to ng dialect.
+        // Simply maps ngraph tensors to values and generate an OP. No op-specific logic.
+        template <typename Op>
+        mlir::Operation* createGenericOp(const ngraph::Node* ngNode);
+
+        template <typename RedOp>
+        mlir::Operation* createIndexReduction(const ngraph::Node* ngNode);
+
+        void createReturn();
+
+        /// Converts nGraph shape-like types \p ng_shape to MLIR shape \p mlir_shape.
+        template <typename T>
+        void getMlirShape(T ngShape, llvm::SmallVectorImpl<int64_t>& mlirShape);
+
+        /// Converts an ngraph shape to an I64 array attribute
+        template <typename T>
+        mlir::ArrayAttr getShapeAsAttr(T ngShape);
+
+    private:
+        // Sub-graph to be compiled and executed with MLIR.
+        const ngraph::op::CompiledKernel* m_compiledKernel;
+        MLIRCompiler* m_compiler;
```

Comment:
Do we need m_compiler in this Pass?

Reply:
its only to get the `module` from the compiler. I was not aware of the fact just `get_module()`  should give the module owned by the compiler https://github.com/NervanaSystems/ngraph/blob/master/src/contrib/mlir/core/compiler.hpp#L136. Will check this out